### PR TITLE
Add display of internal circuits for ControlFlowOps to mpl circuit drawer

### PR DESCRIPTION
**Pull request by enavarro51**
### Summary

This PR adds display of the circuits inside ForLoopOps and SwitchCaseOps to the mpl circuit drawer

### Details and comments

This PR should be merged after #10096 and #10170. This adds the display of ForLoopOps and SwitchCaseOps in the same manner as was done in #10170 for IfElseOps and WhileLoopOps. This also fixes a bug in how gates were folding introduced in #10170.

- [x] Add tests
- [x] Release note

```
qr = QuantumRegister(4, "q")
cr = ClassicalRegister(3, "cr")
qc = QuantumCircuit(qr, cr)
qc.h(0)
with qc.for_loop((2, 4, 8, 16)) as i:
    qc.h(0)
    qc.h(1)
    qc.cx(0, 1)
    qc.rx(pi/i, 1)
    qc.measure(0, 0)
    qc.measure(1, 1)
    with qc.if_test((cr[2], 1)):
        qc.z(0)
qc.draw(style={"showindex": True})
```
image
```
from qiskit.circuit import QuantumCircuit, ClassicalRegister, QuantumRegister
qreg = QuantumRegister(3)
creg = ClassicalRegister(3)
qc = QuantumCircuit(qreg, creg)
qc.h([0, 1, 2])
qc.measure([0, 1, 2], [0, 1, 2])

with qc.switch(creg) as case:
    with case(0):
        qc.x(0)
    with case(1):
        qc.z(1)
        qc.z(1)
    with case(2, 3, 4, 5, 6):
        qc.y(0)
        qc.y(0)
        qc.y(0)
    with case(case.DEFAULT):
        qc.cx(0, 1)
qc.h(0)
qc.draw(style={"showindex": True})
```
image

### CODE REVIEW

File: qiskit/visualization/circuit/matplotlib.py [hunk 1]
```diff
@@ -441,6 +476,100 @@ def _get_layer_widths(self, node_data, glob_data):
...
+
+                        # Recursively call _get_layer_widths for the circuit inside the ControlFlowOp
+                        flow_widths = flow_drawer._get_layer_widths(node_data, wire_map, glob_data)
+                        layer_widths.update(flow_widths)
+
+                        # Gates within a SwitchCaseOp need to know which case they are in
+                        for flow_layer in nodes:
+                            for flow_node in flow_layer:
+                                if isinstance(node.op, SwitchCaseOp):
+                                    node_data[flow_node]["case_num"] = circ_num
```
Comment (jakelishman) [inline qiskit/visualization/circuit/matplotlib.py:557]: How come they need to know?

File: qiskit/visualization/circuit/matplotlib.py [hunk 2]
```diff
@@ -759,11 +946,48 @@ def _draw_regs_wires(self, num_folds, xmax, max_x_index, qubits_dict, clbits_dic
...
+                nodes += flow_drawer[i]._nodes
+                flow_drawer[i]._get_coords(
+                    node_data,
+                    wire_map,
+                    layer_widths,
+                    qubits_dict,
+                    clbits_dict,
+                    glob_data,
+                    flow_parent=flow_drawer[i]._flow_parent,
+                    is_if_while=(i == 0),
```
Comment (jakelishman) [inline qiskit/visualization/circuit/matplotlib.py:967]: What's `is_if_while` actually measuring here?  It looks like it's more like a test of `is_first_block`, and if so, is it not important which number block it is _other_ than "zero" or "not zero"?

**Review by jakelishman — COMMENTED** [review]
Thanks so much for this Edwin.  I've left a couple of minor comments in line, but mostly I'm judging by the test output for this, and it looks fantastic.  I think it might be one of those things where we need to put it out in the wild, and see if other people find any issues with visualisation.

File: qiskit/visualization/circuit/_utils.py [hunk 3]
```diff
@@ -390,7 +390,10 @@ def _get_layered_instructions(
     # default to left
     justify = justify if justify in ("right", "none") else "left"
 
-    qubits = circuit.qubits.copy()
+    if wire_map is not None:
+        qubits = [bit for bit in list(wire_map.keys()).copy() if isinstance(bit, Qubit)]
+    else:
+        qubits = circuit.qubits.copy()
```
[the remark this answers is not in the document]
Comment (enavarro51) [reply qiskit/visualization/circuit/_utils.py:396]: We don't iterate over `wire_map` so order is not an issue. Will make the change.

File: qiskit/visualization/circuit/matplotlib.py [hunk 4]
```diff
@@ -267,9 +293,9 @@ def draw(self, filename=None, verbose=False):
         # load the wire map
         wire_map = get_wire_map(self._circuit, self._qubits + self._clbits, self._cregbundle)
 
-        # node_data per node with 'width', 'gate_text', 'raw_gate_text',
-        # 'ctrl_text', 'param_text', q_xy', and 'c_xy',
-        # and colors 'fc', 'ec', 'lc', 'sc', 'gt', and 'tc'
+        # node_data per node with "width", "gate_text", "raw_gate_text", "ctrl_text",
+        # "param_text", "if_depth", "inside_flow", "x_index", "indexset", "jump_values",
+        # "case_num", "q_xy", "c_xy", and colors "fc", "ec", "lc", "sc", "gt", and "tc"
         node_data = {}
```
[the remark this answers is not in the document]
Comment (enavarro51) [reply qiskit/visualization/circuit/matplotlib.py:299]: Had the exact same thought recently. Create a `NodeData` class and a `node_data` dict keyed on the nodes. Then all these  become attributes that can be documented. For a new PR.

File: qiskit/visualization/circuit/matplotlib.py [hunk 5]
```diff
@@ -441,6 +476,100 @@ def _get_layer_widths(self, node_data, glob_data):
...
+                        circuit_list.append(circ)
+                    elif isinstance(op, SwitchCaseOp):
+                        node_data[node]["jump_values"] = []
+                        cases = list(op.cases_specifier())
+
+                        # Create an empty circuit for the Switch box
+                        circuit_list.append(cases[0][1].copy_empty_like())
+                        for jump_values, circ in cases:
+                            node_data[node]["jump_values"].append(jump_values)
+                            circuit_list.append(circ)
```
[the remark this answers is not in the document]
Comment (enavarro51) [reply qiskit/visualization/circuit/matplotlib.py:506]: I'll test this and see if it works and will then see if we can eliminate the None checks for Else.

File: qiskit/visualization/circuit/matplotlib.py [hunk 6]
```diff
@@ -441,6 +476,100 @@ def _get_layer_widths(self, node_data, glob_data):
...
+                            node_data[node]["if_depth"] = (
+                                node_data[self._flow_parent]["if_depth"] + 1
+                            )
+                        # Update the wire_map with the qubits from the inner circuit
+                        inner_wire_map = {
+                            inner: wire_map[outer]
+                            for outer, inner in zip(self._qubits, circuit.qubits)
+                            if inner not in wire_map
+                        }
+                        wire_map.update(inner_wire_map)
```
[the remark this answers is not in the document]
Comment (enavarro51) [reply qiskit/visualization/circuit/matplotlib.py:527]: Ah, I see. So the concern is that if this is done for an `If` that adds bits, then later there's a `While` that adds other bits, `wire_map `would contain all of those and there could be conflicts. I'll look at what it takes to only use `inner_wire_map`.

File: qiskit/visualization/circuit/matplotlib.py [hunk 7]
```diff
@@ -441,6 +476,100 @@ def _get_layer_widths(self, node_data, glob_data):
...
+                        if circuit is None:  # If with no else
+                            self._flow_drawers[node].append(None)
+                            node_data[node]["width"].append(0.0)
+                            break
+
+                        # Depth of nested ControlFlowOp used for color of box
+                        if self._flow_parent is not None:
+                            node_data[node]["if_depth"] = (
+                                node_data[self._flow_parent]["if_depth"] + 1
+                            )
```
[the remark this answers is not in the document]
Comment (enavarro51) [reply qiskit/visualization/circuit/matplotlib.py:520]: Yeah, I called this `nest_depth` in the text drawer and that's more appropriate.

File: qiskit/visualization/circuit/matplotlib.py (hunk 1 above)
Comment (enavarro51) [reply qiskit/visualization/circuit/matplotlib.py:557]: Check lines 699-704. They have to know which cases are to the left for width calcs.

File: qiskit/visualization/circuit/matplotlib.py (hunk 2 above)
Comment (enavarro51) [reply qiskit/visualization/circuit/matplotlib.py:967]: Right. This is used in `get_coords` to do the extra width calcs for Else and Case. At that point in the code we just need to know that we're working with a circuit that's in an Else or Case and not If, While, For, or Switch. So `is_first_block` is a better name or `is_not_first_block` to go the other way.

File: qiskit/visualization/circuit/matplotlib.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jakelishman) [reply qiskit/visualization/circuit/matplotlib.py:527]: Yeah, precisely.  It's not that users _should_ build circuits that have those sorts of issues, but more that it's possible and there's nothing technically wrong with them doing it by the normal Qiskit rules.

File: qiskit/visualization/circuit/matplotlib.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (enavarro51) [reply qiskit/visualization/circuit/matplotlib.py]: Done in https://github.com/Qiskit/qiskit-terra/pull/10207/commits/8d232e92c08e286f6ee30c1621243446e7f126e2.

File: qiskit/visualization/circuit/matplotlib.py [hunk 8]
```diff
@@ -441,6 +476,100 @@ def _get_layer_widths(self, node_data, glob_data):
...
+                            node_data[node]["jump_values"].append(jump_values)
+                            circuit_list.append(circ)
+
+                    # Now process the circuits inside the ControlFlowOps
+                    for circ_num, circuit in enumerate(circuit_list):
+                        raw_gate_width = 0.0
+                        if circuit is None:  # If with no else
+                            self._flow_drawers[node].append(None)
+                            node_data[node]["width"].append(0.0)
+                            break
```
[the remark this answers is not in the document]
Comment (enavarro51) [reply qiskit/visualization/circuit/matplotlib.py]: Done in https://github.com/Qiskit/qiskit-terra/pull/10207/commits/8d232e92c08e286f6ee30c1621243446e7f126e2.

File: qiskit/visualization/circuit/matplotlib.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (enavarro51) [reply qiskit/visualization/circuit/matplotlib.py]: Done in https://github.com/Qiskit/qiskit-terra/pull/10207/commits/8d232e92c08e286f6ee30c1621243446e7f126e2.

File: qiskit/visualization/circuit/matplotlib.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (enavarro51) [reply qiskit/visualization/circuit/matplotlib.py]: Done in https://github.com/Qiskit/qiskit-terra/pull/10207/commits/8d232e92c08e286f6ee30c1621243446e7f126e2.

File: qiskit/visualization/circuit/matplotlib.py (hunk 2 above)
Comment (enavarro51) [reply qiskit/visualization/circuit/matplotlib.py]: Done in https://github.com/Qiskit/qiskit-terra/pull/10207/commits/8d232e92c08e286f6ee30c1621243446e7f126e2.

**Comment by enavarro51** [comment]
I added a test using `textbook` style in order to test the bug I found with text color.

**Review by jakelishman — APPROVED** [review]
Thanks so much for this, Edwin - I know it was a huge amount of work, both in design and actual execution, and I really appreciate it!  These look fantastic, and I'm really keen to get them into the hands of our users, so we've got better ways to actually understand what's going on in a dynamic circuit.

**Pull request merged**